From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Bj=C3=B6rn=20Bidar?= <bjorn.bidar@jolla.com>
Date: Tue, 31 Aug 2021 01:34:24 +0300
Subject: [PATCH] Fix shlib.lds generation

Generate sed command line used to create shlib.lds before invoking it
to prevent buffer overflow in make.

Fixes MER#472
---
 Makerules | 23 ++++++++++++-----------
 1 file changed, 12 insertions(+), 11 deletions(-)

diff --git a/Makerules b/Makerules
index 83bdd3a44d0d42db6634b91f26bc1d126aa43c3f..6b85423c4a12bed4518688ac55f9cae3857117d5 100644
--- a/Makerules
+++ b/Makerules
@@ -553,25 +553,19 @@ shlib-lds-flags =
 else
 # binutils only position loadable notes into the first page for binaries,
 # not for shared objects
-$(common-objpfx)shlib.lds: $(common-objpfx)config.make $(..)Makerules
-	$(LINK.o) -shared -Wl,-O1 \
-		  -nostdlib -nostartfiles \
-		  $(sysdep-LDFLAGS) $(rtld-LDFLAGS) $(LDFLAGS.so) \
-		  -Wl,--verbose 2>&1 | \
-	  sed > $@T \
-	      -e '/^=========/,/^=========/!d;/^=========/d' \
+shlib-lds-hash = -e '/^=========/,/^=========/!d;/^=========/d' \
 	      $(if $(filter yes,$(have-hash-style)), \
 		   -e 's/^.*\.gnu\.hash[ 	]*:.*$$/  .note.ABI-tag : { *(.note.ABI-tag) } &/' \
 		   -e '/^[ 	]*\.hash[ 	]*:.*$$/{h;d;}' \
 		   -e '/DATA_SEGMENT_ALIGN/{H;g}' \
 		, \
 		   -e 's/^.*\.hash[ 	]*:.*$$/  .note.ABI-tag : { *(.note.ABI-tag) } &/' \
-	       ) \
-	      -e 's/^.*\*(\.dynbss).*$$/& \
+	       )
+shlib-lds-dynbss = -e 's/^.*\*(\.dynbss).*$$/& \
 		 PROVIDE(__start___libc_freeres_ptrs = .); \
 		 *(__libc_freeres_ptrs) \
-		 PROVIDE(__stop___libc_freeres_ptrs = .);/'\
-	      -e 's@^.*\*(\.jcr).*$$@& \
+		 PROVIDE(__stop___libc_freeres_ptrs = .);/'
+shlib-lds-jcr = -e 's@^.*\*(\.jcr).*$$@& \
 		 PROVIDE(__start___libc_subfreeres = .);\
 		 __libc_subfreeres : { *(__libc_subfreeres) }\
 		 PROVIDE(__stop___libc_subfreeres = .);\
@@ -582,6 +576,13 @@ $(common-objpfx)shlib.lds: $(common-objpfx)config.make $(..)Makerules
 		 __libc_IO_vtables : { *(__libc_IO_vtables) }\
 		 PROVIDE(__stop___libc_IO_vtables = .);\
 		 /DISCARD/ : { *(.gnu.glibc-stub.*) }@'
+$(common-objpfx)shlib.lds: $(common-objpfx)config.make $(..)Makerules
+	$(LINK.o) -shared -Wl,-O1 \
+		  -nostdlib -nostartfiles \
+		  $(sysdep-LDFLAGS) $(rtld-LDFLAGS) $(LDFLAGS.so) \
+		  -Wl,--verbose 2>&1 | \
+	  sed > $@T \
+	      $(shlib-lds-hash) $(shlib-lds-dynbss) $(shlib-lds-jcr)
 	test -s $@T
 	mv -f $@T $@
 common-generated += shlib.lds
